pkg/endpoint/endpoint.go

```diff
@@ -2608,3 +2659,71 @@ func (e *Endpoint) syncPolicyMapController() {
 func (e *Endpoint) IsDisconnecting() bool {
 	return e.state == StateDisconnected || e.state == StateDisconnecting
 }
+
+// Lock returns error if endpoint was removed, locks underlying mutex otherwise
+func (e *Endpoint) Lock() error {
+	if e.IsDisconnecting() {
+		return fmt.Errorf("Endpoint was removed before lock attempt.")
+	}
+	e.mutex.Lock()
```

Comment:
I think that this should occur before the `e.IsDisconnecting()` call accesses `e.state` without holding a lock.

Reply:
Can you point me to this call? This method should not be called without error handling, which should abort all accesses, since lock is not acquired.

Reply:
@joestringer ^^

Reply:
The comment above `e.IsDisconnecting()` describes how it must be called while holding `e.mutex.Lock()` to ensure safe access; otherwise there is no guarantee that by the time the call to `IsDisconnecting()` returns that the state hasn't changed.

`IsDisconnecting()` itself reads `e.state`, which should be protected by the endpoint Lock.

In this new Lock() call, `e.state` is being checked without grabbing the lock. So, it seems plausible that this branch statement is checked from one goroutine, then some other code updates the state, then this code grabs the lock. In that race condition, `e.Lock()` will return success with a locked endpoint in disconnecting state.

Reply:
Ah, thanks, I missed this.

Reply:
Done